### PR TITLE
Add FORM

### PR DESCRIPTION
**Pull request by iarspider**
Split from  #19394

### CODE REVIEW

File: var/spack/repos/builtin/packages/form/package.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    url      = "https://github.com/vermaseren/form/archive/v4.2.1.tar.gz"
+
+    version('4.2.1', sha256='6f32c7470d00e8ab6934dc352f5a78e29290146a00e5775f8cd5fef7810bbbb8')
+
+    depends_on('autoconf', type='build')
+    depends_on('automake', type='build')
+    depends_on('libtool',  type='build')
+    depends_on('m4',       type='build')
+    depends_on('gmp', type=('build', 'run'))
+    depends_on('zlib', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/form/package.py]: These two dependencies don't have any binaries, so they can't be used at build-time or run-time. Did you mean `type='link'`?

File: var/spack/repos/builtin/packages/form/package.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+    def autoreconf(self, spec, prefix):
+        autoreconf('--install', '--verbose', '--force')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/form/package.py]: Does the default implementation not work? We basically run this but with additional m4 dirs.

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/form/package.py (hunk 1 above)
Comment (iarspider) [reply var/spack/repos/builtin/packages/form/package.py]: For some reason I though `link` is a part of `build` and `run`. Fixed.

File: var/spack/repos/builtin/packages/form/package.py (hunk 2 above)
Comment (iarspider) [reply var/spack/repos/builtin/packages/form/package.py]: Fixed

**Review by adamjstewart — APPROVED** [review]

**Comment by adamjstewart** [comment]
If you're ever confused about deptypes, I've tried to make https://spack.readthedocs.io/en/latest/packaging_guide.html#dependency-types as clear as possible, but if you notice anything unclear let me know and I can add to the docs.

**Pull request merged**